vfio/src/vfio_device.rs

```diff
@@ -0,0 +1,594 @@
+        Ok(VfioContainer { container })
+    }
+
+    fn get_api_version(&self) -> i32 {
+        // Safe as file is vfio container fd and ioctl is defined by kernel.
+        unsafe { ioctl(self, VFIO_GET_API_VERSION()) }
+    }
+
+    fn check_extension(&self, val: u32) -> Result<()> {
+        if val != VFIO_TYPE1_IOMMU && val != VFIO_TYPE1v2_IOMMU {
+            return Err(VfioError::VfioInvalidType);
+        }
+
+        // Safe as file is vfio container and make sure val is valid.
+        let ret = unsafe { ioctl_with_val(self, VFIO_CHECK_EXTENSION(), val.into()) };
+        if ret != 1 {
+            return Err(VfioError::VfioExtension);
+        }
+
+        Ok(())
+    }
+
+    fn set_iommu(&self, val: u32) -> Result<()> {
+        if val != VFIO_TYPE1_IOMMU && val != VFIO_TYPE1v2_IOMMU {
+            return Err(VfioError::VfioInvalidType);
+        }
+
+        // Safe as file is vfio container and make sure val is valid.
+        let ret = unsafe { ioctl_with_val(self, VFIO_SET_IOMMU(), val.into()) };
+        if ret < 0 {
+            return Err(VfioError::ContainerSetIOMMU);
+        }
+
+        Ok(())
+    }
+
+    fn vfio_dma_map(&self, iova: u64, size: u64, user_addr: u64) -> Result<()> {
+        let dma_map = vfio_iommu_type1_dma_map {
+            argsz: mem::size_of::<vfio_iommu_type1_dma_map>() as u32,
+            flags: VFIO_DMA_MAP_FLAG_READ | VFIO_DMA_MAP_FLAG_WRITE,
+            vaddr: user_addr,
+            iova,
+            size,
+        };
+
+        // Safe as file is vfio container, dma_map is constructed by us, and
+        // we check the return value
+        let ret = unsafe { ioctl_with_ref(self, VFIO_IOMMU_MAP_DMA(), &dma_map) };
+        if ret != 0 {
+            return Err(VfioError::IommuDmaMap);
+        }
+
+        Ok(())
+    }
+
+    fn vfio_dma_unmap(&self, iova: u64, size: u64) -> Result<()> {
+        let mut dma_unmap = vfio_iommu_type1_dma_unmap {
+            argsz: mem::size_of::<vfio_iommu_type1_dma_unmap>() as u32,
+            flags: 0,
+            iova,
+            size,
+        };
+
+        // Safe as file is vfio container, dma_unmap is constructed by us, and
+        // we check the return value
+        let ret = unsafe { ioctl_with_mut_ref(self, VFIO_IOMMU_UNMAP_DMA(), &mut dma_unmap) };
+        if ret != 0 || dma_unmap.size != size {
+            return Err(VfioError::IommuDmaUnmap);
+        }
+
+        Ok(())
+    }
+}
+
+impl AsRawFd for VfioContainer {
+    fn as_raw_fd(&self) -> RawFd {
+        self.container.as_raw_fd()
+    }
+}
+
+struct VfioGroup {
+    group: File,
+    device: DeviceFd,
+    container: VfioContainer,
+}
+
+impl VfioGroup {
+    fn new(id: u32, vm: &VmFd) -> Result<Self> {
+        let mut group_path = String::from("/dev/vfio/");
+        let s_id = &id;
+        group_path.push_str(s_id.to_string().as_str());
+
+        let group = OpenOptions::new()
+            .read(true)
+            .write(true)
+            .open(Path::new(&group_path))
```

Comment:
It seems a little odd you do the path manipulation above using a raw string then create the Path here.

What do you think about the following instead?
```let group_path = Path::new("/dev/vfio").join(id.to_string())```

Reply:
Thanks for the suggestion, it looks nicer. Applied.